src/hooks/useQueryParam/index.ts

```diff
@@ -0,0 +1,54 @@
+import { useCallback } from "react";
+import { useNavigate, useSearchParams } from "react-router-dom";
+import { parseQueryString, stringifyQuery } from "utils/query-string";
+
+/** `useQueryParams` returns all of the query params passed into the url */
+const useQueryParams = () => {
+  const [searchParams] = useSearchParams();
+  const navigate = useNavigate();
+  const setQueryString = useCallback(
+    (params: { [key: string]: any }) => {
+      const stringifiedQuery = stringifyQuery(params);
+      navigate(`?${stringifiedQuery}`, { replace: true });
+    },
+    [navigate]
+  );
+
+  return [parseQueryString(searchParams.toString()), setQueryString] as const;
+};
+
+/** `useQueryParam` allows you to interact with a query param in the same way you would use a useState hook.  */
+const useQueryParam = <T>(
+  param: string,
+  def: T
+): readonly [T, (set: T) => void] => {
+  const [searchParams, setSearchParams] = useQueryParams();
+  const setQueryParam = useCallback(
+    (value: T) => {
+      setSearchParams({
+        ...searchParams,
+        [param]: value,
+      });
+    },
+    [setSearchParams, searchParams, param]
+  );
+
+  return [
+    searchParams[param] !== undefined
+      ? (conditionalToArray(
+          searchParams[param],
+          Array.isArray(def)
+        ) as unknown as T)
+      : def,
+    setQueryParam,
+  ] as const;
+};
+
+/** `conditionalToArray` takes in a value and transforms it into an array if it is not one and should be */
+const conditionalToArray = <T>(value: T, shouldBeArray: boolean) => {
```

Comment:
Is `shouldBeArray` superfluous? It seems like in the usage above you make a call to `Array.isArray()` twice, and I feel like the naming/design of this function overall means that we are assuming that an array will be returned. Would this get the job done?

```js
const conditionalToArray = <T>(value: T) => Array.isArray(value) ? value : [value];
```

Reply:
not necessarily if our default type is an array we are able to determine that this type should always be an array and if we should return an array. 

The reason this is necessary is that `query-param` does not know if a value is an array unless it has 2 or more values. 
so 

`?someArray=1`
This would always return the value `1` instead of `[1]`

`?someArray=1,2`
This would return `[1,2]` 

By adding the shouldBeArray flag we can programmatically determine if our resulting value should be an array because the default value was an array. We can't only rely on just the current value.